src/aleph/http/client.clj

```diff
@@ -498,401 +957,157 @@
-                              (or ssl-context
-                                  (if insecure?
-                                    (netty/insecure-ssl-client-context)
-                                    (netty/ssl-client-context))))
-       :bootstrap-transform bootstrap-transform
-       :remote-address      (InetSocketAddress.
-                             (.getHost uri)
-                             (int
-                              (if (neg? (.getPort uri))
-                                (if ssl? 443 80)
-                                (.getPort uri))))
-       :local-address       local-address
-       :transport           (netty/determine-transport transport epoll?)})
-     (fn [_] s))))
+        authority (str host (when explicit-port? (str ":" port)))
+
+        desired-protocols (mapv {:http1 ApplicationProtocolNames/HTTP_1_1
+                                 :http2 ApplicationProtocolNames/HTTP_2}
+                                http-versions)
+
+        ssl-context (client-ssl-context ssl? ssl-context desired-protocols insecure?)
+
+        logger (cond
+                 (instance? LoggingHandler log-activity) log-activity
+                 (some? log-activity) (netty/activity-logger "aleph-client" log-activity)
+                 :else nil)
+
+        pipeline-builder (make-pipeline-builder
+                           (assoc opts
+                                  :ssl? ssl?
+                                  :ssl-context ssl-context
+                                  :remote-address remote-address
+                                  :raw-stream? raw-stream?
+                                  :response-buffer-size response-buffer-size
+                                  :logger logger
+                                  :pipeline-transform pipeline-transform))
+
+        ch (netty/create-client-chan
+             {:pipeline-builder    pipeline-builder
+              :bootstrap-transform bootstrap-transform
+              :remote-address      remote-address
+              :local-address       local-address
+              :transport           (netty/determine-transport transport epoll?)
+              :name-resolver       name-resolver})]
+
+    (d/chain' ch
+              (fn setup-client
+                [^Channel ch]
+                (log/debug "Channel:" ch)
+
+                ;; We know the SSL handshake must be complete because create-client wraps the
+                ;; future with maybe-ssl-handshake-future, so we can get the negotiated
+                ;; protocol, falling back to HTTP/1.1 by default.
+                (let [pipeline (.pipeline ch)
+                      protocol (if ssl?
+                                 (or (-> pipeline
+                                         ^SslHandler (.get ^Class SslHandler)
+                                         (.applicationProtocol))
+                                     ApplicationProtocolNames/HTTP_1_1)
+                                 ApplicationProtocolNames/HTTP_1_1)
```

Comment:
Can you confirm we are using HTTP1.1 because of this? https://httpwg.org/specs/rfc7540.html#TLSUsage

Reply:
I'm not sure I follow. If you're referring to the final clause of the `or`, we set it to HTTP/1.1 because that's Aleph's default if we're not using SSL/TLS.

The other use of ApplicationProtocolNames/HTTP_1_1, on line 1029, is a fallback in case we ARE using SSL, but the server is NOT using ALPN for some reason. Since HTTP/2 requires ALPN, we shouldn't let that be the default. (And any server old enough to not use ALPN is probably defaulting to HTTP/1 anyway.)

(BTW, RFC 7540 is obsoleted by RFC 9113.)

Reply:
Thanks for the comment :
> ; Not using ALPN, HTTP/2 isn't allowed